packages/application-generic/src/services/in-memory-provider/cache-in-memory-provider.service.ts

```diff
@@ -0,0 +1,105 @@
+import { Logger } from '@nestjs/common';
+
+import { InMemoryProviderService } from './in-memory-provider.service';
+import {
+  InMemoryProviderEnum,
+  InMemoryProviderClient,
+  ScanStream,
+} from './types';
+
+import { GetIsInMemoryClusterModeEnabled } from '../../usecases';
+
+const LOG_CONTEXT = 'CacheInMemoryProviderService';
+
+export class CacheInMemoryProviderService {
+  public inMemoryProviderService: InMemoryProviderService;
+  public isCluster: boolean;
+  private getIsInMemoryClusterModeEnabled: GetIsInMemoryClusterModeEnabled;
+
+  constructor() {
+    this.getIsInMemoryClusterModeEnabled =
+      new GetIsInMemoryClusterModeEnabled();
+
+    const provider = this.selectProvider();
+    this.isCluster = this.isClusterMode();
+
+    const enableAutoPipelining =
+      process.env.REDIS_CACHE_ENABLE_AUTOPIPELINING === 'true';
+
+    this.inMemoryProviderService = new InMemoryProviderService(
+      provider,
+      this.isCluster,
+      enableAutoPipelining
+    );
+  }
+
+  /**
+   * Rules for the provider selection:
+   * - For our self hosted users we assume all of them have a single node Redis
+   * instance.
+   * - For Novu we will use Elasticache. We fallback to a Redis Cluster configuration
+   * if Elasticache not configured properly. That's happening in the provider
+   * mapping in the /in-memory-provider/providers/index.ts
+   */
+  private selectProvider(): InMemoryProviderEnum {
+    if (process.env.IS_DOCKER_HOSTED) {
+      return InMemoryProviderEnum.REDIS;
+    }
+
+    return InMemoryProviderEnum.ELASTICACHE;
+  }
+
+  private descriptiveLogMessage(message) {
+    return `[Provider: ${this.selectProvider()}] ${message}`;
+  }
+
+  private isClusterMode(): boolean {
+    const isClusterModeEnabled = this.getIsInMemoryClusterModeEnabled.execute();
+
+    Logger.log(
+      this.descriptiveLogMessage(
+        `Cluster mode ${
+          isClusterModeEnabled ? 'is' : 'is not'
```

Comment:
Nit for readability on the differentiator.

```suggestion
          isClusterModeEnabled ? 'IS' : 'IS NOT'
```